controllers/node_label_module_version_reconciler.go

```diff
@@ -0,0 +1,257 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/constants"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/filter"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/utils"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+//+kubebuilder:rbac:groups="core",resources=nodes,verbs=get;watch;patch
+
+// this struct contains all the version labels related to a specific Module
+type modulesVersionLabels struct {
+	name                     string
+	namespace                string
+	moduleVersionLabel       string
+	moduleLoaderVersionLabel string
+	devicePluginVersionLabel string
+}
+
+const (
+	NodeLabelModuleVersionReconcilerName = "NodeLabelModuleVersion"
+)
+
+type reconcileLabelsResult struct {
+	labelsToAdd    map[string]string
+	labelsToDelete []string
+	requeue        bool
+}
+
+type NodeLabelModuleVersionReconciler struct {
+	client    client.Client
+	helperAPI nodeLabelModuleVersionHelperAPI
+}
+
+func NewNodeLabelModuleVersionReconciler(client client.Client) *NodeLabelModuleVersionReconciler {
+	return &NodeLabelModuleVersionReconciler{
+		client:    client,
+		helperAPI: newNodeLabelModuleVersionHelper(client),
+	}
+}
+
+func (nlmvr *NodeLabelModuleVersionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	node := v1.Node{}
+
+	if err := nlmvr.client.Get(ctx, types.NamespacedName{Name: req.Name}, &node); err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not get node: %v", err)
+	}
+
+	logger := log.FromContext(ctx).WithValues("node name", node.Name)
+
+	// get node's label data per module
+	modulesVersionLabels := nlmvr.helperAPI.getLabelsPerModules(node.Labels)
+
+	// get module loader pods on the node
+	moduleLoaderPods, err := nlmvr.helperAPI.getModuleLoaderPods(ctx, node.Name)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could get module loader pods for the node %s: %v", node.Name, err)
+	}
+
+	// get the labels to add and delete for this specific node
```

Comment:
Can we remove this comment, the code is self explanatory.

Reply:
fixed